dtm0/log.c

```diff
@@ -30,24 +30,495 @@
+	m0_sm_group_unlock(grp);
+	return M0_RC(0);
+}
+
+M0_INTERNAL void m0_dtm0_log_destroy(struct m0_dtm0_log *dol)
+{
+	/* XXX No-op for now. TODO implement */
+}
+
+static int dtm0_log0_init(struct m0_be_domain *dom, const char *suffix,
+                          const struct m0_buf *data)
+{
+	return 0;
+}
+
+static void dtm0_log0_fini(struct m0_be_domain *dom, const char *suffix,
+                           const struct m0_buf *data)
+{
+}
+
+static inline m0_bcount_t dtm0_log_transactions_ksize(const void *key)
+{
+	return sizeof ((struct dtm0_log_record *)NULL)->lr_descriptor.dtd_id;
+}
+
+static inline m0_bcount_t dtm0_log_transactions_vsize(const void *value)
+{
+	return sizeof(struct dtm0_log_record *);
+}
+
+static int dtm0_log_transactions_compare(const void *key0, const void *key1)
+{
+	return m0_dtx0_id_cmp(key0, key1);
+}
+
+static int redo_insert_callback(struct m0_btree_cb  *cb,
+				 struct m0_btree_rec *rec)
+{
+	struct m0_btree_rec     *datum = cb->c_datum;
+
+	/** Write the Key and Value to the location indicated in rec. */
+	m0_bufvec_copy(&rec->r_key.k_data,  &datum->r_key.k_data,
+		       m0_vec_count(&datum->r_key.k_data.ov_vec));
+	m0_bufvec_copy(&rec->r_val, &datum->r_val,
+		       m0_vec_count(&rec->r_val.ov_vec));
+	return 0;
+}
+
+static int redo_insert(struct m0_btree *tree, struct m0_be_tx *tx,
+			     struct m0_buf *key, struct m0_buf *val)
+{
+	struct m0_btree_op   kv_op        = {};
+	void                *k_ptr = key->b_addr;
+	void                *v_ptr = val->b_addr;
+	m0_bcount_t          ksize = key->b_nob;
+	m0_bcount_t          vsize = val->b_nob;
+	struct m0_btree_rec  rec = {
+			   .r_key.k_data = M0_BUFVEC_INIT_BUF(&k_ptr, &ksize),
+			   .r_val        = M0_BUFVEC_INIT_BUF(&v_ptr, &vsize),
+			   .r_crc_type   = M0_BCT_NO_CRC,
+			};
+	struct m0_btree_cb   redo_insert_cb = {.c_act = redo_insert_callback,
+					  .c_datum = &rec,
+					 };
+	return M0_BTREE_OP_SYNC_WITH_RC(&kv_op,
+					m0_btree_put(tree, &rec, &redo_insert_cb,
+						     &kv_op, tx));
+}
+
+M0_INTERNAL int m0_dtm0_log_redo_add(struct m0_dtm0_log        *dol,
+                                     struct m0_be_tx           *tx,
+                                     const struct m0_dtm0_redo *redo,
+                                     const struct m0_fid       *p_sdev_fid)
+{
+	struct dtm0_log_record  *rec;
+	struct m0_be_seg        *seg = dol->dtl_cfg.dlc_seg;
+
+	M0_LOG(M0_DEBUG, "fid:"FID_F, FID_P(p_sdev_fid));
+	m0_mutex_lock(&dol->dtl_lock);
+	M0_PRE(!dol->dtl_the_end);
+	M0_PRE(dtm0_log_invariant(dol));
+	m0_mutex_unlock(&dol->dtl_lock);
+
+	/* TODO lookup before insert */
+	/* TODO check memory allocation errors */
+	M0_BE_ALLOC_PTR_SYNC(rec, seg, tx);
+	M0_ASSERT(redo->dtr_payload.dtp_data.ab_count == 1);
+	rec->lr_payload_data.b_nob =
+		redo->dtr_payload.dtp_data.ab_elems[0].b_nob;
+	M0_BE_ALLOC_BUF_SYNC(&rec->lr_payload_data, seg, tx);
+	m0_buf_memcpy(&rec->lr_payload_data,
+		      &redo->dtr_payload.dtp_data.ab_elems[0]);
```

Comment:
As a performance optimization, we need to achieve ZERO-COPY.

Reply:
Due to be list link, i don't see a way to avoid it at this moment.
 struct m0_be_list_link    lr_link_all_p;
 
 If we don't need this all_p list, we can avoid this BE allocation and directly use the redo structure.

Reply:
Added a todo for it.